### ISSUE TITLE
Queue Processing Of Credit Memos Causes Doubled Order Refund Totals

### ISSUE THREAD

**Issue by james481** [classyllama/ClassyLlama_AvaTax#29]
On Magento EE 2.0.7 / module version 0.2.4, processing the Avatax queue with online captured credit memos results in Magento incorrectly incrementing the refunded totals on the linked order when the credit memo is saved again after processing, resulting in orders that show doubled refund amounts.

This appears to be an issue with the core Magento code in `Magento\Sales\Model\ResourceModel\Order\Creditmemo\Relation\Refund` that does not expect the refunded credit memo to be saved multiple times (and doesn't check if it should add the refunded amount to the order refunded totals again). We've filed a support issue with Magento, but it's not clear if they will choose to change this behavior in the core (and when that may be released), so we decided to file this issue as well so that in the event it's not fixed in core the module logic can adjust the credit memo save relation hooks to account for this issue.

**Comment by erikhansen** [comment]
@james481 Thanks for reporting this issue. When testing this extension, we didn't test with online credit memos, so we wouldn't have seen this issue in testing.

The reason we save the credit memo is because the module adds two columns to the credit memo record that we update with the AvaTax response from the AvaTax API:   `avatax_is_unbalanced` and `base_avatax_tax_amount`. Those values are purely for debuggin/informational purposes.

Let me know what Magento support says.

**Comment by erikhansen** [comment]
@james481 Did you hear back from Magento about this issue?

**Comment by erikhansen** [comment]
@james481 Ping.

**Comment by casconed** [comment]
I'm seeing this problem as well.

**Comment by casconed** [comment]
Got this response from Magento Support:
>After working with my engineer on this we determined our native offering is working as it was designed to work. We do understand that ava tax is trying to save the credit memo again. From my standpoint logically once a credit memo has been done for a full refund there should be no other reason to do another credit memo which is where the problem creates. 

>Now, maybe from a business standpoint we may have determined to accommodate this behavior and add the code to our core to allow this. So my engineer is doing some work on this including looking for the ticket classy lama opened with us regarding this. I will update you when we have determined what is going on with this. Thanks for your patience while this is looked into further.

**Comment by mttjohnson** [comment]
The AvaTax module doesn't save the credit memo a second time, it uses a plugin to wrap around the save method that is called by anything else and inserts a couple additional fields into the credit memo (avatax_is_unbalanced, base_avatax_tax_amount).

https://github.com/classyllama/ClassyLlama_AvaTax/blob/develop/Plugin/Sales/Model/Spi/CreditmemoResource.php

**Comment by sunilit42** [comment]
Hello Team, AvaTax Module doing credit Memo second time.

Let me explain how module doing.

When we generate credit memo in magento, magento capture amount,  your module submit request to avatex for credit back that amount.

That time it is called below function
ClassyLlama\AvaTax\Model\Queue\Task.php File
Function name processPendingQueue()
and it is called 
ClassyLlama\AvaTax\Model\Queue\Processing.php
There is one 
 $this->updateAdditionalEntityAttributes($entity, $processSalesResponse);
from there you are updating attribute into credit memo model and it will call your 
ClassyLlama\AvaTax\Plugin\Sales\Model\Spi\CreditmemoResource  and it is capture payment second time offline.
So it is calling second times.

So basically what you need to do, you need to modify updateAdditionalEntityAttributes function, i think you need to write custom query and do direct update into table, no need to save model again.

**Comment by james481** [comment]
Sorry for my absence @erikhansen, this got lost in my notification stream somewhere. I wasn't successful in getting any further information from Magento support on this issue, as they required a database dump of the affected site which I wasn't able to provide before proceeding on further debugging. I did work directly with one of the Magento platform architects while diagnosing it, and he confirmed the issue was created by the way Magento handles the state check on incrementing the refunded order amounts, but I don't have any further information on any plans for this to be fixed in the core (though the above seems to indicate it might not be).

**Comment by erikhansen** [comment]
@james481 @casconed I just merged in PR #41 to the develop branch, which fixes this issue. If you need the fix immediately, you can switch to using the develop branch. Over the course of the next week, we'll be releasing this fix in a new release.

**Comment by mttjohnson** [comment]
Just in case anyone isn't familiar with how to use the develop branch, you can specify the branch name with the composer require command
`composer require classyllama/module-avatax:dev-develop`

That is the main difference related to the installation instructions
http://www.classyllama.com/documentation/extensions/avatax-magento-2-module
